### PR TITLE
ENH: Dependent censoring

### PR DESCRIPTION
**Pull request by kshedden**
This provides methods for estimating marginal survival and cumulative incidence functions when there is dependent censoring.  The user must provide covariates such that the censoring becomes independent conditioned on the covariates.

The primary reference is:

http://arxiv.org/pdf/math/0409180.pdf

### CODE REVIEW

**Comment by josef-pkt** [comment]
based on a quick read-through:
I didn't see any problems

n -> nobs for number of observations, avoids having to look up what n is
the kernel module should get a leading underscore to privatize it, given that there are no public functions in it.

Aside: whenever I see kernel function, I wonder whether splines or series estimators would also be possible, and faster than the kernel approach because they don't need the nobs loop. In that case, calling the supporting module `_nonparametric` would be more general.

**Comment by kshedden** [comment]
In this case the kernel is used for matching not basis expansion.  It is somewhat expensive but not as bad as I thought (can easily do thousands to low 10's of thousands).

The fast alternative for big data sets is to stratify on the variables used to do the kernel matching, then run standard KM or CumIncidence within each stratum, then average.  That runs in seconds on n=60K.

The kernel method should be more efficient for smaller data sets (n ~100s) where stratification is not practical.

File: statsmodels/duration/survfunc.py [hunk 1]
```diff
@@ -154,7 +169,18 @@ class CumIncidenceRight(object):
         cinc[k-1] contains the estimated cumulative incidence rates
         for outcome k=1,2,...
     cinc_se : list of arrays
-        The standard errors for the values in `cinc`.
+        The standard errors for the values in `cinc`.  Not available when
+        exog and/or frequency weights are provided.
+
+    Notes
+    -----
+    When exog is provded, a local estimate of the cumulative incidence
```
Comment (josef-pkt) [inline statsmodels/duration/survfunc.py]: Typo provided missing i

File: statsmodels/duration/survfunc.py [hunk 2]
```diff
@@ -206,31 +251,68 @@ class SurvfuncRight(object):
         Optional title used for plots and summary output.
     freq_weights : array-like
         Optional frequency weights
+    exog : array-like
+        Optional, if present used to account for violation of
+        independent censoring.
+    bwm : float
+        Band-width multiplier for kernel-baed estimation.  Only
```
Comment (josef-pkt) [inline statsmodels/duration/survfunc.py]: Typo kernel-based missing s

**Comment by kshedden** [comment]
Vector `bwm` seems to work, test added.

**Comment by kshedden** [comment]
Now rescaling the data internally prior to kernel fitting so that the bandwidth has a fixed meaning.

**Comment by kshedden** [comment]
Suggestions on names are welcome...

**Comment by josef-pkt** [comment]
What's the status of this? Did we already merge some related code or is this still open and missing?

**Comment by kshedden** [comment]
It's still open.  I will rebase and we can see if it is ready to merge.

**Comment by josef-pkt** [comment]
Another skimming of the changes.
Looks good overall, maybe a few names are still too short, e.g. bwm -> bw_factor ?

If I have time tomorrow, then I would like to get a rough idea how the method works, otherwise we merge as is, and I figure this out later. 
My impression is that eventually we will get a reusable pattern out of this and similar model or case specific methods.

File: statsmodels/duration/_kernel_estimates.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    Parameters
+    ----------
+    time : array-like
+        The observed times for each subject
+    status : array-like
+        The status for each subject (1 indicates event, 0 indicates
+        censoring)
+    exog : array-like
+        Covariates such that censoring is independnet conditioned on
```
Comment (josef-pkt) [inline statsmodels/duration/_kernel_estimates.py]: typo "independent"
and I think it should be "conditional"

File: statsmodels/duration/_kernel_estimates.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    probs : array-like
+        The estimated survival probabilities
+    times : array-like
+        The times at which the survival probabilities are estimated
+
+    References
+    ----------
+    Estimating marginal survival function by adjusting for dependent
+    censoring using many covariates
+    http://arxiv.org/pdf/math/0409180.pdf
```
Comment (josef-pkt) [inline statsmodels/duration/_kernel_estimates.py]: full citation

Zeng, Donglin. 2004. “Estimating Marginal Survival Function by Adjusting for Dependent Censoring Using Many Covariates.” The Annals of Statistics 32 (4): 1533–55. doi:10.1214/009053604000000508.

**Comment by josef-pkt** [comment]
I read the main part of the Zeng 2004 article, and added the `comp-treatment` label because of the similarity to the doubly robust estimators in the treatment effect literature.
There is no treatment variable here, and it is more `missing at random` with conditional independence but we don't have a specific label for that.
nonparametric: This uses something like conditional kernel density estimation instead of inverse probability weighting to condition out the endogenous "drop-out".
The kernel is hardcoded to Gaussian equivalent.

There are interesting connection to those other parts, but, IMO, too early to think about code reuse and generic solutions. (Also, I skipped the treatment in survival models literature so far.)

I found a typo, but otherwise you could hit the merge button.

File: statsmodels/duration/_kernel_estimates.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        The estimated survival probabilities
+    times : array-like
+        The times at which the survival probabilities are estimated
+
+    References
+    ----------
+    Zeng, Donglin 2004. Estimating Marginal Survival Function by
+    Adjusting for Dependent Censoring Using Many Covariates. The
+    Annals of Statistics 32 (4):
+    1533–55. doi:10.1214/009053604000000508.
```
Comment (josef-pkt) [inline statsmodels/duration/_kernel_estimates.py]: the dash `-` in the number `1533–55` is not ASCII
appveyor failes

**Pull request merged**